### PR TITLE
Introduce lambda based timeout assertions

### PR DESCRIPTION
**Pull request by sbrannen**
## Overview

This commit introduces `assertTimeout()` methods in `Assertions` that support both lazy and preemptive timeouts for blocks of code supplied as lambda expressions.

For example, the following assertion would pass:

`assertTimeout(ofMillis(500), () -> Thread.sleep(10));`

Whereas, the following assertion would fail:

`assertTimeout(ofMillis(10), () -> Thread.sleep(500));`

The above examples execute the code block in the same thread as the caller. To perform the same assertion as the last example and have the code executed in a separate thread and then preemptively aborted, the above can be rewritten as follows.

`assertTimeoutPreemptively(ofMillis(10), () -> Thread.sleep(500));`

Issue: #438

---

I hereby agree to the terms of the JUnit Contributor License Agreement.

### CODE REVIEW

**Comment by sbrannen** [comment]
@junit-team/junit-lambda, please let me know your thoughts on the PR.

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/Assertions.java [hunk 1]
```diff
@@ -996,6 +1005,154 @@ public static void assertThrows(Class<? extends Throwable> expectedType, Executa
 			String.format("Expected %s to be thrown, but nothing was thrown.", expectedType.getName()));
 	}
 
+	// --- assert timeouts -----------------------------------------------
+
+	/**
+	 * <em>Asserts</em> that execution of the supplied {@code executable}
+	 * completes before the given {@code timeout} is exceeded.
+	 *
+	 * <p>Note: the executable will be executed in the same thread as	 that
```
Comment (ttddyy) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/Assertions.java]: nit: space between "as" and "that". same for other timeout methods

Comment (sbrannen) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/Assertions.java]: good catch! (y)

I'll fix it.

**Comment by marcphilipp** [comment]
Looks good to me! Great to finally have support for timeouts! 👍

**Comment by sbrannen** [comment]
Thanks for the review, and... I couldn't agree more: it definitely is great to finally have official support for timeouts in JUnit 5!

**Pull request merged**